ert_storage/endpoints/blob_handler.py

```diff
@@ -0,0 +1,297 @@
+import io
+from typing import Optional, List, AsyncGenerator
+from uuid import uuid4, UUID
+
+import numpy as np
+import pandas as pd
+from fastapi import (
+    Request,
+    UploadFile,
+)
+from fastapi.logger import logger
+from fastapi.responses import Response, StreamingResponse
+
+from ert_storage import database_schema as ds
+from ert_storage.database import Session, HAS_AZURE_BLOB_STORAGE
+
+if HAS_AZURE_BLOB_STORAGE:
+    from ert_storage.database import azure_blob_container
+
+
+class GeneralBlobHandler:
+    def __init__(
+        self,
+        db: Session,
+        name: Optional[str],
+        ensemble_id: Optional[UUID],
+        realization_index: Optional[int],
+    ):
+        self._db = db
+        self._name = name
+        self._ensemble_id = ensemble_id
+        self._realization_index = realization_index
+
+    async def upload_blob(
+        self,
+        file: UploadFile,
+    ) -> ds.File:
+        file_obj = ds.File(
+            filename=file.filename,
+            mimetype=file.content_type,
+        )
+        file_obj.content = await file.read()
+
+        return file_obj
+
+    async def stage_blob(
+        self,
+        request: Request,
+        block_index: int,
+    ) -> ds.FileBlock:
+        ensemble = self._db.query(ds.Ensemble).filter_by(id=self._ensemble_id).one()
+        block_id = str(uuid4())
+
+        file_block_obj = ds.FileBlock(
+            ensemble=ensemble,
+            block_id=block_id,
+            block_index=block_index,
+            record_name=self._name,
+            realization_index=self._realization_index,
+        )
+        file_block_obj.content = await request.body()
+
+        return file_block_obj
+
+    def create_blob(self) -> ds.File:
+        file_obj = ds.File(
+            filename="test",
+            mimetype="mime/type",
+        )
+
+        return file_obj
+
+    async def finalize_blob(self) -> None:
+        ensemble = self._db.query(ds.Ensemble).filter_by(id=self._ensemble_id).one()
+
+        record_obj = self._get_record(
+            ensemble_pk=ensemble.pk,
+        )
+
+        submitted_blocks = self._get_submitted_blocks(
+            ensemble_pk=ensemble.pk,
+        )
+
+        data = b"".join([block.content for block in submitted_blocks])
+        record_obj.file.content = data
+
+    async def extract_content(
+        self, record: ds.Record, accept: Optional[str], realization_index: Optional[int]
+    ) -> Response:
+        type_ = record.record_info.record_type
+        if type_ == ds.RecordType.f64_matrix:
```

Comment:
This is a little confusing to me. Blobs are records of type `file` . The matrices are not considered blobs.

Reply:
I know, it is kind of confusing. I was actually thinking of renaming the class to GeneralContentHandler or something? I started off refactoring the handling of the blobs only, but then all of it was moved into this class eventually because it needed the same structure. But i'm not sure if this is the correct path at all, i'm not totally convinced my solution is THE solution.... @dotfloat will make an attempt taking my concept a bit further, we'll see how it goes.

Reply:
I think it'd make sense to have two helper classes, one for blobs (which automatically uses Azure if available) and one for matrices, which can interpret csv and json. But all in all I think a confusing class is better than what we have now.